velox/functions/prestosql/ArrayFunctions.h

```diff
@@ -316,4 +320,89 @@ struct ArraySumFunction {
   }
 };
 
+/// Function Signature: array_duplicates(array(T)) -> array(T)
+/// where T must be coercible to bigint or varchar.
+/// Returns a set of elements that occur more than once in array.
+///
+/// Internally, we maintain a hashmap to indicate non-null tri-state:
+/// 1. key does not exist (key not in map),
+/// 2. unique (map-value=true),
+/// 3. duplicate (map-value=false).
+/// Meanwhile, maintain a std::optional to indicate null tri-state:
+/// 1. null does not exist (std::nullopt)
+/// 2. unique (value=true),
+/// 3. duplicate (value=false).
+template <typename TExecCtx, typename T>
+struct ArrayDuplicatesFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(TExecCtx);
+
+  FOLLY_ALWAYS_INLINE void call(
+      out_type<velox::Array<T>>& out,
+      const arg_type<velox::Array<T>>& inputArray) {
+    folly::F14FastMap<arg_type<T>, bool> uniqueMap;
+    std::optional<bool> nullState = std::nullopt;
+    // TODO: Introducing an ordered set is redundant
+    //  if std::sort() in ArrayWriter is implemented
+    std::set<arg_type<T>> duplicateNonNullElements;
+
+    for (const auto& item : inputArray) {
+      if (item.has_value()) {
+        auto it = uniqueMap.find(*item);
+        if (it == uniqueMap.end()) {
+          uniqueMap[*item] = true;
+        } else if (it->second) {
+          it->second = false;
+          duplicateNonNullElements.insert(*item);
+        }
+      } else {
+        if (nullState.has_value()) {
+          if (nullState.value()) {
+            nullState = false;
+          }
+        } else {
+          nullState = true;
+        }
+      }
+    }
+
+    if (!nullState.value_or(true)) {
+      out.add_null();
+    }
+    for (const auto& item : duplicateNonNullElements) {
+      if constexpr (std::is_same_v<T, Varchar>) {
+        out.add_item().setNoCopy(item);
```

Comment:
Good to know that.
I was told by @laithsakka to use `setNoCopy` for string zero copy optimization from source to result. The simple function doc mentioned:
```
StringWriter<>
When Zero-copy optimization is enabled (see zero-copy-string-result section above), the following functions can be used.
- void setEmpty(): set to empty string.
- void setNoCopy(const StringView& value): set string to an input string without performing deep copy.
```
So, if we aim to enable zero copy optimization from source strings to result, shall I simply add `reuse_strings_from_arg` variable to my class and keep `setNoCopy()` as is?

Reply:
It should be sufficient to add reuse_strings_from_arg variable. However, I don't quite remember whether it works for arrays of strings in addition to just strings. Can you do some debugging to confirm? We need to make sure the result vector has a reference to the string buffers pointed to by the StringViews.

Reply:
Debug result is: `reuse_strings_from_arg` only support `Varchar` type.
Complex types such as `Array` isn't supported yet. Thus, this line: https://github.com/darrenfu/velox/blob/4639baf8042face63bbf46272301b1b311247371/velox/expression/SimpleFunctionAdapter.h#L360-L360 will fail in my case.

Once `reuse_strings_from_arg` supports `Array` type, I will flip the flag value from -1 to 0 to enable zero copy.

Reply:
@darrenfu Thank you for debugging. Are you saying we currently cannot implement this function as a simple function without copying the strings? If that's the case, it seems we should continue using Vector function. 

@laithsakka Laith, what would it take to allow this function to avoid copying strings?

Reply:
> @darrenfu Thank you for debugging. Are you saying we currently cannot implement this function as a simple function without copying the strings? If that's the case, it seems we should continue using Vector function.

Yes, that's correct. But I believe enabling it in simple function api should be a minor change. Let me talk with @laithsakka how to change the simple api.

Reply:
> @darrenfu Thank you for debugging. Are you saying we currently cannot implement this function as a simple function without copying the strings? If that's the case, it seems we should continue using Vector function.
> 
> @laithsakka Laith, what would it take to allow this function to avoid copying strings?

@mbasmanova since this UDF will need to support generic types in the next version of presto (also see issue https://github.com/facebookincubator/velox/issues/3484), I think it's wise to abandon this simple function solution coz we will have to use vector function anyway then. LMK what you think.

Reply:
Makes sense to me.

Reply:
Close this PR given we will still be in favor of the origin vector function implementation.